### DISCUSSION TITLE
How does oven media transcode bitstream which don't have IDR frame

### DISCUSSION THREAD

**Discussion by captain-bugs** [OvenMediaLabs/OvenMediaEngine#1451] [Help]
in wanted to know how part/bitstream is transcoded if there is not IDR frame in it 
in this section
https://github.com/AirenSoft/OvenMediaEngine/blob/ee8ee528bb0ecdba658d235d9d460d76efa99c84/src/projects/mediarouter/mediarouter_stream.cpp#L369 
if SPS and PPS is not present code is adding it before IDR but what if there is no IDR frame, will it still transcode it?

**Comment by getroot** [comment]
A video stream without IDR frames does not exist. That means the stream cannot be decoded.

**Reply by captain-bugs** [reply]
its the part segments i'm talking about, if the part that is supposed to be transcoded does not contain the IDR, in that case how will transcoding take place.
because as per specs every part (considering it 500ms part) in the manifest that has `INDEPENDENT=YES` attribute means it can be played without referring previous frames, which means that part has IDR frame in it.
So is it true that other parts without `INDEPENDENT=YES` attribute don't have IDR frame in it? If yes then how is it able to transcode/decode it?

**Reply by getroot** [reply]
The source code you attached is MediaRouter, which manages OME's internal streams. LLHLS is then packaged by the LLHLS Pulbisher.

Anyway, if that were the question, the player would start playing with INDEPENDENT=YES.

**Reply by captain-bugs** [reply]
1. can you tell why do we need to add SPS/PPS nalu before IDR if not present?
2. also can you point me where the transcoding actually happens? like scaling of framerate change?

**Reply by getroot** [reply]
The short answer is as follows.
1. For WebRTC Playback
2. https://github.com/AirenSoft/OvenMediaEngine/tree/master/src/projects/transcoder

**Reply by getroot** [reply]
I think it would make our discussion go more smoothly if you let me know what you are actually trying to do. Are you just studying by analyzing OvenMediaEngine code?

**Reply by captain-bugs** [reply]
yes, I wan't to do transcoding process in one of my golang project for LHLS, so wanted to understand how transcoding works with NAL units

**Reply by getroot** [reply]
The term transcoding you use seems different from mine. We use the term transcoding to mean decoding and re-encoding. Could you elaborate on what it means to transcode in NAL units? Do you mean mp4 packaging for LLHLS?

**Reply by captain-bugs** [reply]
scaling, changing framerate, adding filtergraph etc.
